[feature](inverted index)WIP inverted index api: SQL syntax and metadata

xiaokang opened the pull request:
# Proposed changes

Issue Number: Step1.1 of [DSIP-023: Add inverted index for full text search](https://cwiki.apache.org/confluence/display/DORIS/DSIP-023%3A+Add+inverted+index+for+full+text+search?src=contextnavpagetreemode)

## Problem summary

Introduce a SQL syntax for creating inverted index and related metadata changes.

```
-- create table with INVERTED index 

CREATE TABLE httplogs (
  ts datetime,
  clientip varchar(20),
  request string,
  status smallint,
  size int,
  INDEX idx_size (size) USING INVERTED,
  INDEX idx_status (status) USING INVERTED,
  INDEX idx_clientip (clientip) USING INVERTED PROPERTIES("parser"="none")
)
DUPLICATE KEY(ts)
DISTRIBUTED BY RANDOM BUCKETS 10


-- add an INVERTED index  to a table

CREATE INDEX idx_request ON httplogs(request) USING INVERTED PROPERTIES("parser"="english");
```

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

xiaokang committed:
WIP inverted index api: SQL syntax and metadata

xiaokang committed:
fix clang format

xiaokang committed:
fix testcase error

xiaokang committed:
fix getInvertedIndexParser and tablet_meta default

xiaokang committed:
remove PARSER_NOT_SET and use PARSER_NONE as default

xiaokang committed:
fix getInvertedIndexParser

morningman commented on fe/fe-core/src/main/cup/sql_parser.cup:
```diff
@@ -3173,6 +3174,10 @@ opt_index_type ::=
     {:
         RESULT = IndexDef.IndexType.BITMAP;
     :}
+    | KW_USING KW_INVERTED
```
missing BLOOM_FILTER?

morningman reviewed (dismissed):
LGTM

xiaokang committed:
index meta refactor:
1. seperate index meta from column meta in be meta
2. use index_id to uniquely identify a index

xiaokang committed:
add index properties and id in show index output

xiaokang committed:
add testcase for index meta

xiaokang committed:
fix clang format

hello-stephen commented:
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.2 seconds
 load time: 572 seconds
 storage size: 17154644867 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221031174821_clickbench_pr_36671.html

xiaokang committed:
fix nullpointer in initIndexUniqueId

morningman commented on gensrc/proto/olap_file.proto:
```diff
@@ -201,6 +201,20 @@ message ColumnPB {
+}
+
+message TabletIndexPB {
+    required int32 index_id = 1;
```
do not use `required`, use `optional`

xiaokang replied:
delete index_name

xiaokang replied:
changed

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/TableIndexes.java:
```diff
@@ -56,6 +58,26 @@ public TableIndexes(List<Index> indexes, Map<String, String> properties) {
         this.properties = properties;
     }
 
+    public int incAndGetMaxIndexUniqueId() {
```
```suggestion
    private int incAndGetMaxIndexUniqueId() {
```

xiaokang replied:
changed

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/TableIndexes.java:
```diff
@@ -56,6 +58,26 @@ public TableIndexes(List<Index> indexes, Map<String, String> properties) {
+        return this.maxIndexId;
+    }
+
+    public int getMaxIndexUniqueId() {
```
```suggestion
    private int getMaxIndexUniqueId() {
```

xiaokang replied:
changed

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/Index.java:
```diff
@@ -30,36 +31,55 @@
+    public static final int INDEX_ID_INIT_VALUE = -1;
+
+    @SerializedName(value = "indexId")
+    private int indexId;
```
How to do the forward compatibility? that the old index does not has an ID.

xiaokang replied:
set default value to -1 for compatibility

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/TableIndexes.java:
```diff
@@ -40,6 +40,8 @@ public class TableIndexes implements Writable {
     @SerializedName(value = "properties")
     private Map<String, String> properties;
+    @SerializedName(value = "maxIndexId")
+    private int maxIndexId = Index.INDEX_ID_INIT_VALUE;
```
Why not using `Env.getNextId()`?

xiaokang replied:
add synchrolized for protection

xiaokang committed:
set indexId to -1 for compatibality

xiaokang committed:
add synchrolized for indexId for concurrency control

xiaokang committed:
change TabletIndexPB fields to optional

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/Index.java:
```diff
@@ -84,6 +104,36 @@ public void setIndexType(IndexDef.IndexType indexType) {
+    }
+
+    public String getPropertiesString() {
+        if (properties == null || properties.isEmpty()) {
```
you can use `PrintableMap` to do this.

morningman commented on fe/fe-core/src/main/java/org/apache/doris/analysis/IndexDef.java:
```diff
@@ -126,29 +153,37 @@ public boolean isSetIfNotExists() {
 
     public void checkColumn(Column column, KeysType keysType) throws AnalysisException {
-        if (indexType == IndexType.BITMAP) {
+        if (indexType == IndexType.BITMAP || indexType == IndexType.INVERTED) {
```
I think we can also add `BLOOM_FILTER` check this time, although it is not implemented yet.

xiaokang replied:
added

morningman commented on fe/fe-core/src/main/java/org/apache/doris/analysis/IndexDef.java:
```diff
@@ -48,12 +52,18 @@ public IndexDef(String indexName, boolean ifNotExists, List<String> columns, Ind
     public void analyze() throws AnalysisException {
-        if (indexType == IndexDef.IndexType.BITMAP) {
+        if (indexType == IndexDef.IndexType.BITMAP
+                || indexType == IndexDef.IndexType.INVERTED) {
```
I think we can also add `BLOOM_FILTER` check this time, although it is not implemented yet.

xiaokang replied:
added

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/TableIndexes.java:
```diff
@@ -56,6 +58,26 @@ public TableIndexes(List<Index> indexes, Map<String, String> properties) {
         this.properties = properties;
     }
 
+    private synchronized int incAndGetMaxIndexUniqueId() {
```
Oh, there is still a problem to use `maxIndexId` inside `TableIndexes`.
For example, after `add index`, there will be an edit log, and after FE restart, it will replay that edit log, but it didn't update the `maxIndexId` in `TableIndexes`.

xiaokang replied:
change to Env.getCurrentEnv().getNextId()

xiaokang committed:
use Env.getCurrentEnv().getNextId() for index_id and add BLOOMFILTER index type

morningman reviewed (approved):
LGTM
